docs/API/metadata.md

```diff
@@ -0,0 +1,43 @@
+
+Clients can also set arbitrary metadata through the "dashboard" and "user" metadata namespaces:
+* The "dashboard" namespace can only be modified by the owner of the dataset, and is visible to anyone with read access to the dataset.
+* The "user" namespace is private to each authenticated user, and even if you don't own a dataset you can set your own private "user" metadata to help you categorize that dataset and to find it again.
+
+Metadata namespaces are hierarchical, and are exposed as nested JSON objects. You can address an entire namespace, e.g., `dashboard` or `dataset` and retrieve the entire JSON object, or you can address nested objects or values using a dotted metadata key path like `dashboard.contact.email` or `dataset.metalog.pbench.script`.
+
+## Metadata namespaces
+
+There are currently three metadata namespaces.
+
+The `dataset` and `server` namespaces are defined and managed by Pbench. The `dashboard` namespace allows an authenticated client to define an arbitrary nested set of JSON objects. The `user` namespace is similar; except that where metadata in the `dashboard` namespace can only be modified by the owner of the dataset, and is visible to all clients with read access to the dataset, any authenticated user can set arbitrary values in the `user` namespace and those values are visible only to the same authenticated user.
+
+### Dataset namespace
+
+This defines the dataset resource, and contains metadata from the dataset tarball recieved from the Pbench agent, including the full contents of a `metadata.log` file which was generated when the agent packaged up the result data into a tarball.
+
+This namespace includes the resource name, which can be modified by the owner of the dataset, but all other metadata in this namespace is controlled by the server.
+
+### Server namespace
+
+This defines internal server state related to a dataset that's not inherent to the representation of the data. Most of this is not of much use to external clients, but can be observed.
+
+Only `server.deletion` can be modified by an authenticated client on behalf of the owner of the dataset. When a dataset is uploaded, the server computes an expected time representing the user's default data retention period based on the time the dataset was uploaded. The owner of the dataset is allowed to modify this later, as long as the new timestamp is within the maximum allowed server data retention period.
+
+### Dashboard namespace
+
+The server will never modify or directly interpret values in this namespace. An authenticated client representing the owner of a dataset can set any keys within this namespace to any desired values for retrieval later. All clients with read access to the dataset will see the same values.
```

Comment:
We need to be careful with "desired values".  Could the APIs be used to built up a huge set of keys and data the consume the database?  We might want to consider some kind of limits on this, configured by the server (total # of keys, total size of values, etc.).

Reply:
Yeah, I assume that there will be a number of places where we want to institute limits to prevent abuse (number of keys, bytes per key, total bytes, etc.).  But, for the moment, I think we're relying on the fact that you have to be a registered user in order to create anything, so that should protect us from the main part of the onslaught.

Reply:
We are not just protecting against the malicious but the inadvertent.